src/util/importJSON.ts

```diff
@@ -167,52 +156,56 @@ const saveThoughts = (
           updatedBy,
         )
 
-        // TODO: remove mutations
-        contextIndexUpdates[contextEncoded] = parent
-        // `Parent` entry for the newly created thought.
-        contextIndexUpdates[newThought.id] = newThought
-
-        thoughtIndexUpdates[hashThought(nonDuplicateValue)] = lexeme
+        return {
+          thoughtIndexUpdates: {
+            [hashThought(nonDuplicateValue)]: lexeme,
+          },
+          contextIndexUpdates: {
+            [contextEncoded]: parent,
+            [newThought.id]: newThought,
+          },
+        }
       }
 
+      const insertUpdates = !skipLevel ? insert() : null
+
+      const updatedState = insertUpdates
+        ? {
+            ...stateNewBeforeInsert,
+            thoughts: {
+              ...stateNewBeforeInsert.thoughts,
+              contextIndex: mergeUpdates(stateNewBeforeInsert.thoughts.contextIndex, insertUpdates.contextIndexUpdates),
+              thoughtIndex: mergeUpdates(stateNewBeforeInsert.thoughts.thoughtIndex, insertUpdates.thoughtIndexUpdates),
+            },
+          }
+        : stateNewBeforeInsert
+
       const updatedDuplicateIndex = {
         ...accum.duplicateIndex,
         [hashedValue]: duplicateValueCount ? duplicateValueCount + 1 : 1,
       }
 
-      const updatedState: State = {
-        ...state,
-        thoughts: {
-          ...state.thoughts,
-          contextIndex: {
-            ...state.thoughts.contextIndex,
-            ...contextIndexUpdates,
-          },
-          thoughtIndex: {
-            ...state.thoughts.thoughtIndex,
-            ...thoughtIndexUpdates,
-          },
-        },
-      }
-
       /**
        *
        */
       const getLastAddedChild = () => {
-        const parent = contextIndexUpdates[contextEncoded]
+        const parent = updatedState.thoughts.contextIndex[contextEncoded]
         return (childIdsToThoughts(updatedState, parent.children) ?? []).find(
           child => child.value === nonDuplicateValue,
         )
       }
 
+      const updatedContextIndexUpdates = mergeUpdates(accum.contextIndex, insertUpdates?.contextIndexUpdates || {})
+      const udpatedThoughtIndexUpdates = mergeUpdates(accum.thoughtIndex, insertUpdates?.thoughtIndexUpdates || {})
+
       const childPath: Path = skipLevel ? path : [...path, getLastAddedChild()!.id]
 
       if (block.children.length > 0) {
         return {
           ...saveThoughts(
             updatedState,
-            contextIndexUpdates,
-            thoughtIndexUpdates,
+            updatedContextIndexUpdates,
+            udpatedThoughtIndexUpdates,
```

Comment:
`saveThoughts` takes `contextIndexUpdates` and `thoughtIndexUpdates`, yet the values passed are full `contextIndex` and `thoughtIndex` types, despite their names, as pointed out above. Did you mean to pass `insertUpdates?.contextIndexUpdates`? Or perhaps aggregate all of the updates? 

If you pass the full `contextIndex` and `thoughtIndex` values, I think it would unintentionally reinsert all thoughts.

Reply:
Actually passing these two parameters is not needed now. I have refactored it to pass the updated state at every level and just return the new added thought indices entries.